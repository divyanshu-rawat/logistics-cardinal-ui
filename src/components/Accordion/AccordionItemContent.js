import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const AccordionItemContentStyled = styled('div')`
  padding: 0 0 20px 0;
`;

const AccordionItemContent = ({ children }) => (
  <AccordionItemContentStyled>{children}</AccordionItemContentStyled>
);

AccordionItemContent.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AccordionItemContent;
