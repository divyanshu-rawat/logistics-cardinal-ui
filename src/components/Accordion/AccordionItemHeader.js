import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Heading from '../Heading';
import Icons from '../Icons';

export const AccordionItemHeaderStyled = styled('div')`
  display: flex;
  align-items: center;
  cursor: pointer;
  height: ${({ theme }) => theme.accordionHeaderHeight};

  &:hover {
    color: ${({ theme }) => theme.primaryColor};
  }

  .AccordionItemHeader {
    &__title {
      flex: 1;
      margin-bottom: 0;
      padding: 0 0 0 2px;
    }
    &__icon {
      padding: 0 10px 0 0;
    }
  }
`;

const AccordionItemHeader = ({ children, action, open }) => (
  <AccordionItemHeaderStyled onClick={action}>
    <Heading as="h3" className="AccordionItemHeader__title">
      {children}
    </Heading>
    <div className="AccordionItemHeader__icon">
      {open ? <Icons.ChevronUp /> : <Icons.ChevronDown />}
    </div>
  </AccordionItemHeaderStyled>
);

AccordionItemHeader.propTypes = {
  children: PropTypes.node.isRequired,
  open: PropTypes.bool.isRequired,
  action: PropTypes.func.isRequired,
};

export default AccordionItemHeader;
