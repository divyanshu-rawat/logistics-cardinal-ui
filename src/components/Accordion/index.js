import React, { Component } from 'react';
import PropTypes from 'prop-types';

import AccordionContext from './Context';

class Accordion extends Component {
  static propTypes = {
    multiple: PropTypes.bool.isRequired,
    children: PropTypes.node.isRequired,
  };

  static defaultProps = {
    multiple: true,
  };

  state = {};

  onToggleItem = (id) => () => {
    const { multiple } = this.props;

    if (!multiple) {
      const entries = Object.keys(this.state);

      this.setState({
        ...entries.reduce(
          (previous, current) => ({ ...previous, [current]: false }),
          {},
        ),
      });
    }

    this.setState({
      [id]: !this.state[id],
    });
  };

  onRegisterComponent = (id, open = false) => {
    this.setState({ [id]: open });
  };

  render() {
    const { children } = this.props;

    return (
      <AccordionContext.Provider
        value={{
          items: this.state,
          toggleItem: this.onToggleItem,
          registerComponent: this.onRegisterComponent,
        }}
      >
        {children}
      </AccordionContext.Provider>
    );
  }
}

export default Accordion;
