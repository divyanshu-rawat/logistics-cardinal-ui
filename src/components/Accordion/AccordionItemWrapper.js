import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Fade from 'react-reveal/Fade';
import AccordionItemHeader from './AccordionItemHeader';
import AccordionItemContent from './AccordionItemContent';

export const AccordionItemDividerStyled = styled('div')`
  border-bottom: 2px solid #2c3037;
`;
class AccordionItemWrapper extends PureComponent {
  static propTypes = {
    registerComponent: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    items: PropTypes.object.isRequired,
    content: PropTypes.node.isRequired,
    title: PropTypes.node.isRequired,
    toggleItem: PropTypes.func.isRequired,
    expandOnInit: PropTypes.bool.isRequired,
  };

  static ANIMATION_DELAY = 400;
  static ANIMATION_DURATION = 700;

  componentDidMount() {
    const { registerComponent, id, expandOnInit } = this.props;
    registerComponent(id, expandOnInit);
  }

  render() {
    const { items, toggleItem, id, title, content } = this.props;
    return (
      <AccordionItemDividerStyled>
        <AccordionItemHeader action={toggleItem(id)} open={!!items[id]}>
          {title}
        </AccordionItemHeader>
        <Fade
          delay={AccordionItemWrapper.ANIMATION_DELAY}
          duration={AccordionItemWrapper.ANIMATION_DURATION}
          collapse
          when={!!items[id]}
        >
          <AccordionItemContent>{content}</AccordionItemContent>
        </Fade>
      </AccordionItemDividerStyled>
    );
  }
}

export default AccordionItemWrapper;
