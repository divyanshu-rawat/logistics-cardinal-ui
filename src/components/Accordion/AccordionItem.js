import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import shortId from 'shortid';

import AccordionItemWrapper from './AccordionItemWrapper';

import AccordionContext from './Context';

class AccordionItem extends PureComponent {
  static propTypes = {
    title: PropTypes.node.isRequired,
    children: PropTypes.node.isRequired,
    expandOnInit: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    expandOnInit: false,
  };

  state = {
    id: shortId.generate(),
  };

  render() {
    const { title, children, expandOnInit } = this.props;
    const { id } = this.state;

    return (
      <AccordionContext.Consumer>
        {({ items, toggleItem, registerComponent }) => (
          <AccordionItemWrapper
            items={items}
            title={title}
            content={children}
            toggleItem={toggleItem}
            registerComponent={registerComponent}
            expandOnInit={expandOnInit}
            id={id}
          />
        )}
      </AccordionContext.Consumer>
    );
  }
}

export default AccordionItem;
