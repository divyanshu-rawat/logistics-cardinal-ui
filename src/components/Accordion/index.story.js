import React from 'react';

import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { select } from '@storybook/addon-knobs';

import withTests from '../../utils/withTests';

import { COMPONENTS } from '../../../.storybook/sections';

import Accordion from '.';
import AccordionItem from './AccordionItem';

storiesOf(`${COMPONENTS}Accordion`, module)
  .addDecorator(
    withTests(
      'Accordion/tests/Accordion',
      'Accordion/tests/AccordionItemContent',
      'Accordion/tests/AccordionItemContentStyled',
      'Accordion/tests/AccordionItemDividerStyled',
      'Accordion/tests/AccordionItemHeader',
      'Accordion/tests/AccordionItemHeaderStyled',
      'Accordion/tests/AccordionItemWrapper',
    ),
  )
  .add(
    'Simple',
    withInfo()(() => (
      <div
        style={{
          background: '#343A42',
          padding: '0 20px',
          width: '500px',
          height: '400px',
        }}
      >
        <Accordion multiple={false}>
          <AccordionItem title="About">
            <p>Some dumb text...</p>
          </AccordionItem>
          <AccordionItem title="Address">
            <p>Some dumb text...</p>
          </AccordionItem>
        </Accordion>
      </div>
    )),
  )
  .add(
    'Allow Multiple',
    withInfo()(() => (
      <div
        style={{
          background: '#343A42',
          padding: '0 20px',
          width: '500px',
          height: '400px',
        }}
      >
        <Accordion>
          <AccordionItem title="About">
            <p>Some dumb text...</p>
          </AccordionItem>
          <AccordionItem title="Address">
            <p>Some dumb text...</p>
          </AccordionItem>
        </Accordion>
      </div>
    )),
  )
  .add(
    'Pre expanded Children',
    withInfo()(() => (
      <div
        style={{
          background: '#343A42',
          padding: '0 20px',
          width: '500px',
          height: '400px',
        }}
      >
        <Accordion>
          <AccordionItem title="About">
            <p>Some dumb text...</p>
          </AccordionItem>
          <AccordionItem title="Address" expandOnInit>
            <p>Some dumb text...</p>
          </AccordionItem>
        </Accordion>
      </div>
    )),
  );
