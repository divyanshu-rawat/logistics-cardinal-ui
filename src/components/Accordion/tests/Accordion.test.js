import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Accordion from '..';
import AccordionItem from '../AccordionItem';

const renderedComponent = ({ multiple = true }) =>
  shallow(
    <Accordion multiple={multiple}>
      <AccordionItem title="Some Title">
        <p>Some Content...</p>
      </AccordionItem>
    </Accordion>,
  );

describe('Accordion', () => {
  it('should render the component', () => {
    const renderComponent = renderedComponent({});
    expect(toJson(renderComponent)).toMatchSnapshot();
  });

  it('onRegisterComponent should register the component with given id', () => {
    const renderComponent = renderedComponent({});
    const KEY = '1bcAdb'; // short-id;
    expect(renderComponent.state(KEY)).toBe(undefined);
    renderComponent.instance().onRegisterComponent(KEY, true);
    expect(renderComponent.state(KEY)).toBe(true);
  });

  it('onToggleItem should toggle component', () => {
    const renderComponent = renderedComponent({});
    const KEY = '1bcAdb'; // short-id;
    const KEY2 = '1bcAdbCD'; // short-id;
    const instance = renderComponent.instance();

    instance.onRegisterComponent(KEY, false);
    instance.onRegisterComponent(KEY2, false);

    instance.onToggleItem(KEY)();

    expect(renderComponent.state(KEY)).toBe(true);

    renderComponent.setProps({ multiple: false });

    instance.onToggleItem(KEY2)();
    expect(renderComponent.state(KEY)).toBe(false);
    expect(renderComponent.state(KEY2)).toBe(true);
  });
});
