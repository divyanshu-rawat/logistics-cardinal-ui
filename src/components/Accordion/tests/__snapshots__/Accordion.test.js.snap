// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "items": Object {},
      "registerComponent": [Function],
      "toggleItem": [Function],
    }
  }
>
  <AccordionItem
    expandOnInit={false}
    title="Some Title"
  >
    <p>
      Some Content...
    </p>
  </AccordionItem>
</ContextProvider>
`;
